client/src/permissions/UserAuth.js

```diff
@@ -0,0 +1,134 @@
+import { UserRoles } from "./UserRoles";
+import "./typedefs"
+import { GlobalContext } from "./contexts/GlobalContext";
+import { TourneyContext } from "./contexts/TourneyContext";
+import { MatchContext } from "./contexts/MatchContext";
+
+export class UserAuth {
+  /**
+   * @type {User}
+   */
+  _user;
+
+  /**
+   * @param {User} user
+   */
+  constructor(user) {
+    this._user = user;
+  }
+
+  /**
+   * @returns {User}
+   */
+  getUser() {
+    return this._user;
+  }
+
+  /**
+   * @param {string} context 
+   * @returns 
+   */
+  withContext(context) {
+    return new UserAuthWithContext(this, context);
+  }
+
+  forGlobal() {
+    return this.withContext(new GlobalContext());
+  }
+
+  /**
+   * @param {string} tourney 
+   * @returns {UserAuthWithContext}
+   */
+  forTourney(tourney) {
+    return this.withContext(new TourneyContext(tourney));
+  }
+
+  /**
+   * @param {Match} match 
+   * @param {number | undefined} playerNo
+   * @param {Map<string, Team> | Team[]} teams
+   * @returns {UserAuthWithContext}
+   */
+  forMatch(match, playerNo, teams) {
+    return this.withContext(new MatchContext(match, playerNo, teams));
+  }
+
+  /**
+   * @param {Team} match 
+   * @returns {UserAuthWithContext}
+   */
+  forTeam(team) {
+    return this.withContext(new TeamContext(team));
+  }
+}
+
+export class UserAuthWithContext {
```

Comment:
Hmm would probably need to refactor since `forTourney()` etc. is defined inside `UserAuth` and we probably don't wanna repeat these in `UserAuthWithContext`

Reply:
Oops didn't see the comment about extending `UserAuth`, applied in aadf9c8079ddac9be16c62ed9a599af60bb059aa